sunscreen_tfhe/gpu_src/src/math/fft/negacyclic.cuh

```diff
@@ -49,31 +49,33 @@ __device__ void twisted_fft(
 /// This result is still in floating point and needs to be modulo reduced.
 template <typename Complex>
 __device__ void twisted_ifft(
-    const Complex * __restrict__ s_input,
+    Complex * __restrict__ s_input,
     typename ScalarOf<Complex>::Ty * __restrict__ s_output,
     uint32_t n)
 {
-    using S = ScalarOf<Complex>::Ty;
+    using S = typename ScalarOf<Complex>::Ty;
     uint32_t n_div_2 = n / 2;
 
     // Perform an n/2 IFFT.
     ifft(s_input, n_div_2);
 
-    S n_inv = Float<S>::TWO / (S)n;
+    S n_inv = 1.0 / (S)n_div_2;
 
     // Twist the inputs so we can use an N-point FFT for negacyclic
     // convolution
-    for (uint32_t i = threadIdx.x; i < n; i += blockDim.x)
+    for (uint32_t i = threadIdx.x; i < n_div_2; i += blockDim.x)
     {
         Complex twist_inv;
-        sincos(-PI * (S)i / (S)n, &twist.y, &twist.x);
+        sincos(-PI * (S)i / (S)n, &twist_inv.y, &twist_inv.x);
```

Comment:
I measured the error of this computation to be up to 2 ULPs. It's interesting this cascades to something like 1e-10 of error.